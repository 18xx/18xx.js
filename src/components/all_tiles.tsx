import * as Immutable from 'immutable';
import { List } from 'immutable';
import * as React from 'react';
import { ReactElement } from 'react';

import TileDefinition, { TileDefinitionInput } from '../tile_definition';

import * as allTilesJson from '../../config/tiles.json';

const allTiles: List<TileDefinitionInput> =
  Immutable.fromJS(allTilesJson).map(
    (el: any) => el.toJS() as TileDefinitionInput
  );

interface AllTilesProps {
  readonly orientation: string;
}

export default class AllTiles
extends React.Component<AllTilesProps, undefined> {

  public get tiles(): any {
    return allTiles.map(t =>
      <div key={t.num}>
        {new TileDefinition(t, this.props.orientation).allRotations}
      </div>
    );
  }

  public render(): ReactElement<AllTiles> {
    return <div id='all-tiles'>{this.tiles}</div>;
  }
}
