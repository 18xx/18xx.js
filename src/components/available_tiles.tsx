import { List, Map } from 'immutable';
import * as _ from 'lodash';
import * as React from 'react';
import { MouseEventHandler, ReactElement } from 'react';

import Tile from './tile';

import TileDefinition from '../tile_definition';
import TileSet from '../tile_set';

export interface AvailableTilesProps {
  readonly tileFilter?: List<string>;
  readonly onClick: any;
  readonly show: boolean;
  readonly tiles: Map<string, string>;
  readonly tileSet: TileSet;
}

export default class AvailableTiles
  extends React.Component<AvailableTilesProps, undefined> {

  public static defaultProps: any;

  get tiles(): List<ReactElement<any>> {
    return this.props.tileSet.all.filter(tileDef =>
      this.props.tileFilter.includes(tileDef.num)
    ).map(
      tileDef => tileDef.allRotations.map(
        (tile: ReactElement<Tile>, index: number) =>
          this.drawTile(tile, tileDef.num, index)
      )
    ).flatten().toList();
  }

  public render(): ReactElement<AvailableTiles> {
    return <div id='tileMenu'>{this.tiles}</div>;
  }

  private drawTile(
    tile: ReactElement<Tile>,
    num: string,
    index: number,
  ): ReactElement<any> {
    const total: number = this.props.tileSet.totalTiles(num);

    let onClick: MouseEventHandler<HTMLElement> = () =>
      this.props.onClick(tile);
    let divClass: string = 'available-tile';
    let availableElement: ReactElement<HTMLElement>;

    if (total) {
      const available: number = total - this.usedTiles(num);

      if (available <= 0) {
        onClick = null;
        divClass += ' unavailable';
      }

      availableElement = (
        <div className='amount-available'>
          {available} / {total}
        </div>
      );
    }

    // FIXME: Move onClick to tile
    return <div
      key={`${num}:${index}`}
      onClick={onClick}
      className={divClass}
      >
        {tile}
        {availableElement}
      </div>;
  }

  private usedTiles(num: string): any {
    return _.countBy(
      this.props.tiles.valueSeq().map((s: string) => s.split('.')[0]).toJS()
    )[num] || 0;
  }
}
