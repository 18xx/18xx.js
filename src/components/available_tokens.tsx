import { List } from 'immutable';
import * as React from 'react';
import { ReactElement } from 'react';

import Token from './token';

import Company from '../company';

export interface AvailableTokensProps {
  readonly companies: List<Company>;
  readonly onClick: (company: Company) => void;
}

class AvailableTokens extends React.Component<AvailableTokensProps, {}> {

  get tokens(): List<ReactElement<Token>> {
    return this.props.companies.map(company =>
      <span
      key={company.shorthand}
      onClick={ () => this.props.onClick(company) }>
        <Token
        text={company.shorthand}
        primaryColor={company.primaryColor}
        secondaryColor={company.secondaryColor}
        textColor={company.textColor} />
      </span>
    ).toList();
  }

  public render(): ReactElement<AvailableTokens> {
    return <div id='tokenMenu'>{this.tokens}</div>;
  }
}

export default AvailableTokens;
