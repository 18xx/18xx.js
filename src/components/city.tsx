import { List } from 'immutable';
import * as _ from 'lodash';
import * as React from 'react';
import { ReactElement } from 'react';

import CityCircle from './city_circle';
import Tile from './tile';
import Token from './token';
import Town from './town';

import CityCircleFactory from '../city_circle_factory';
import Company from '../company';
import Hexagon from '../hexagon';
import Point from '../point';

export interface CityProps {
  hex?: string;
  hexagon: Hexagon;
  onRightClickCity: Function;
  onRightClickToken: Function;
  num: number;
  points?: List<Point>;
  rotation?: number;
  spotLocations?: number[]; // FIXME: Refactor out to distinct city
  tokenState: List<string>;
  homeTokens: List<string>;
}

export interface Station {
  render(): ReactElement<Station>;
}

export default class City
extends React.Component<CityProps, undefined>
implements Station {
  public static defaultProps: CityProps = {
    homeTokens: List<string>(),
    tokenState: List<string>(),
  } as CityProps;

  public render(): ReactElement<City> {
    const num: number = this.props.num;
    let result: ReactElement<any>;

    if (num === 1) {
      result = (
        this.buildCircle(0, this.hexagon.center)
      );
    } else if (num === 2) {
      const background: ReactElement<any> = (
        <rect
          x={this.hexagon.center.x - CityCircle.DEFAULT_RADIUS}
          y={this.hexagon.center.y - CityCircle.DEFAULT_RADIUS -
            CityCircle.STROKE_WIDTH / 2}
          width={CityCircle.DEFAULT_RADIUS * 2}
          height={CityCircle.DEFAULT_RADIUS * 2 + CityCircle.STROKE_WIDTH}
          fill='black'
        />
      );
      const circles: List<ReactElement<CityCircle>> = List([
        this.buildCircle(
          0,
          new Point(
            this.hexagon.center.x - CityCircle.DEFAULT_RADIUS,
            this.hexagon.center.y
          )
        ),
        this.buildCircle(
          1,
          new Point(
            this.hexagon.center.x + CityCircle.DEFAULT_RADIUS,
            this.hexagon.center.y
          )
        )
      ]);
      result = (
        <g key='cities'>
          {background}
          {circles}
        </g>
      );
    } else if (num === 3) {
      const hexPoints: List<Point> = List([
        new Point(
          this.hexagon.center.x - CityCircle.DEFAULT_RADIUS + 8,
          this.hexagon.center.y - 2 * CityCircle.DEFAULT_RADIUS
        ),
        new Point(
          this.hexagon.center.x - 2 * CityCircle.DEFAULT_RADIUS + 1,
          this.hexagon.center.y - CityCircle.DEFAULT_RADIUS
        ),
        new Point(
          this.hexagon.center.x - 2 * CityCircle.DEFAULT_RADIUS + 1,
          this.hexagon.center.y + CityCircle.DEFAULT_RADIUS
        ),
        new Point(
          this.hexagon.center.x - CityCircle.DEFAULT_RADIUS + 8,
          this.hexagon.center.y + 2 * CityCircle.DEFAULT_RADIUS
        ),
        new Point(
          this.hexagon.center.x + CityCircle.DEFAULT_RADIUS + 4,
          this.hexagon.center.y + CityCircle.DEFAULT_RADIUS
        ),
        new Point(
          this.hexagon.center.x + CityCircle.DEFAULT_RADIUS + 4,
          this.hexagon.center.y - CityCircle.DEFAULT_RADIUS
        ),
      ]);

      const background: ReactElement<any> = (
        <polygon points={hexPoints.join(' ')} fill='black' />
      );

      const circles: List<ReactElement<CityCircle>> = List([
        this.buildCircle(
          0,
          new Point(
            this.hexagon.center.x - CityCircle.DEFAULT_RADIUS + 2,
            this.hexagon.center.y - CityCircle.DEFAULT_RADIUS
          )
        ),
        this.buildCircle(
          1,
          new Point(
            this.hexagon.center.x - CityCircle.DEFAULT_RADIUS + 2,
            this.hexagon.center.y + CityCircle.DEFAULT_RADIUS
          )
        ),
        this.buildCircle(
          2,
          new Point(
            this.hexagon.center.x + CityCircle.DEFAULT_RADIUS - 2,
            this.hexagon.center.y
          )
        ),
      ]);

      result = (
        <g key='cities'>
          {background}
          {circles}
        </g>
      );
    } else if (num === 4) {
      const hexPoints: List<Point> = List([
        new Point(
          this.hexagon.center.x - CityCircle.DEFAULT_RADIUS,
          this.hexagon.center.y - 2 * CityCircle.DEFAULT_RADIUS
        ),
        new Point(
          this.hexagon.center.x - 2 * CityCircle.DEFAULT_RADIUS,
          this.hexagon.center.y - CityCircle.DEFAULT_RADIUS
        ),
        new Point(
          this.hexagon.center.x - 2 * CityCircle.DEFAULT_RADIUS,
          this.hexagon.center.y + CityCircle.DEFAULT_RADIUS
        ),
        new Point(
          this.hexagon.center.x - CityCircle.DEFAULT_RADIUS,
          this.hexagon.center.y + 2 * CityCircle.DEFAULT_RADIUS
        ),
        new Point(
          this.hexagon.center.x + CityCircle.DEFAULT_RADIUS,
          this.hexagon.center.y + 2 * CityCircle.DEFAULT_RADIUS
        ),
        new Point(
          this.hexagon.center.x + 2 * CityCircle.DEFAULT_RADIUS,
          this.hexagon.center.y + CityCircle.DEFAULT_RADIUS
        ),
        new Point(
          this.hexagon.center.x + 2 * CityCircle.DEFAULT_RADIUS,
          this.hexagon.center.y - CityCircle.DEFAULT_RADIUS
        ),
        new Point(
          this.hexagon.center.x + CityCircle.DEFAULT_RADIUS,
          this.hexagon.center.y - 2 * CityCircle.DEFAULT_RADIUS
        ),
      ]);

      const background: ReactElement<any> = (
        <polygon points={hexPoints.join(' ')} fill='black' />
      );

      const circles: List<ReactElement<CityCircle>> = List([
        this.buildCircle(
          0,
          new Point(
            this.hexagon.center.x - CityCircle.DEFAULT_RADIUS,
            this.hexagon.center.y - CityCircle.DEFAULT_RADIUS
          )
        ),
        this.buildCircle(
          1,
          new Point(
            this.hexagon.center.x - CityCircle.DEFAULT_RADIUS,
            this.hexagon.center.y + CityCircle.DEFAULT_RADIUS
          )
        ),
        this.buildCircle(
          2,
          new Point(
            this.hexagon.center.x + CityCircle.DEFAULT_RADIUS,
            this.hexagon.center.y - CityCircle.DEFAULT_RADIUS
          )
        ),
        this.buildCircle(
          3,
          new Point(
            this.hexagon.center.x + CityCircle.DEFAULT_RADIUS,
            this.hexagon.center.y + CityCircle.DEFAULT_RADIUS
          )
        ),
      ]);

      result = (
        <g key='cities'>
          {background}
          {circles}
        </g>
      );
    } else if (num === 6) {
      const points: Point[] = _.times(6).map(n =>
        Point.from(this.hexagon.center, n, this.cityCircleRadius * 2),
      );
      const r: number = this.cityCircleRadius + CityCircle.STROKE_WIDTH / 2;

      const bgPoints: Point[] = _.flatten(
        _.times(6).map(n =>
          [
            Point.from(points[n], n - 0.5, r),
            Point.from(points[n], n + 0.5, r),
          ]
        )
      );

      const background: ReactElement<any> = (
        <polygon points={bgPoints.join(' ')} fill='black' />
      );

      const circles: List<ReactElement<CityCircle>> = List(
        points.map((point, idx) => this.buildCircle(idx, point))
      );

      result = (
        <g key='cities'>
          {background}
          {circles}
        </g>
      );
    } else {
      throw new Error('Unsupported number of cities');
    }

    return result;
  }

  protected get cityCircleRadius(): number {
    let result: number = CityCircle.DEFAULT_RADIUS;
    if (this.props.num === 6) {
      result -= 2;
    }
    return result;
  }

  protected buildCircle(index: number, point: Point): ReactElement<CityCircle> {
    const factory: CityCircleFactory = new CityCircleFactory(
      this.props.hex,
      this.props.homeTokens,
      this.props.onRightClickCity,
      this.props.onRightClickToken,
      this.props.tokenState,
      this.cityCircleRadius,
    );
    return factory.build(index, point);
  }

  protected get hexagon(): Hexagon {
    return this.props.hexagon;
  }
}
