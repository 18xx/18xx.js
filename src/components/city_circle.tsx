import * as React from 'react';
import { ReactElement } from 'react';

import { MapHexElement } from './map_hex';
import { TileElement } from './tile';
import Token from './token';

import Point from '../point';

const DEFAULT_RADIUS: number = 19;
const STROKE_WIDTH: number = 2;

export interface CityCircleProps {
  readonly key?: string;
  readonly hex?: string;
  readonly onContextMenu?: any;
  readonly point: Point;
  readonly radius?: number;
  readonly token?: ReactElement<Token>;
}

class CityCircle extends React.Component<CityCircleProps, {}>
implements MapHexElement, TileElement {
  public static defaultProps: Partial<CityCircleProps> = {
    radius: DEFAULT_RADIUS,
  };

  static get DEFAULT_RADIUS(): number {
    return DEFAULT_RADIUS;
  }

  static get STROKE_WIDTH(): number {
    return STROKE_WIDTH;
  }

  get radius(): number {
    return this.props.radius!;
  }

  get point(): Point {
    return this.props.point;
  }

  public render(): ReactElement<CityCircle> {
    const r: number = this.radius + STROKE_WIDTH / 2;
    return (
      <svg
      x={this.point.x - r}
      y={this.point.y - r}
      onContextMenu={ this.props.onContextMenu }>
        <circle
          cx={r}
          cy={r}
          r={this.radius}
          fill='white'
          stroke='#000'
          strokeWidth={STROKE_WIDTH}
          className='city-circle'
        />
        {this.props.token}
      </svg>
    );
  }
}

export default CityCircle;
