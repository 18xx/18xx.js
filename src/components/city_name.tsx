import * as React from 'react';
import { Component, ReactElement } from 'react';

import { MapHexElement } from './map_hex';

import Point from '../point';

interface CityNameProps {
  readonly name: string;
  readonly point: Point;
}

class CityName extends Component<CityNameProps, {}> implements MapHexElement {
  public render(): ReactElement<CityName> {
    return (
      <text
        textAnchor='middle'
        fontSize={this.fontSize}
        x={this.props.point.x}
        y={this.textY}>
        {this.props.name}
      </text>
    );
  }

  private get fontSize(): number {
    let result: number = 16;
    if (this.props.name.length > 12) {
      result = 12;
    }
    return result;
  }

  private get textY(): number {
    return this.props.point.y - 16 + this.fontSize;
  }
}

export default CityName;
