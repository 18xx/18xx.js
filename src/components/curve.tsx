import * as React from 'react';
import { ReactElement } from 'react';

import { TileElement } from './tile';

import Point from '../point';

interface CurveProps {
  readonly point1: Point;
  readonly point2: Point;
  readonly radius: number;
  readonly stroke?: string;
  readonly strokeDasharray?: string;
  readonly strokeWidth?: number;
}

class Curve extends React.Component<CurveProps, {}> implements TileElement {
  public static defaultProps: Partial<CurveProps> = {
    stroke: 'black',
    strokeDasharray: '1,0',
    strokeWidth: 8,
  };

  public render(): ReactElement<Curve> {
    const dValues: string = [
      'M' + this.props.point1.x,
      this.props.point1.y,
      'A',
      this.props.radius,
      this.props.radius,
      0,
      '0,0',
      this.props.point2.x,
      this.props.point2.y,
    ].map(val => val.toString()).join(' ');

    return (
      <path
        d={dValues}
        fill='none'
        stroke={this.props.stroke}
        strokeDasharray={this.props.strokeDasharray}
        strokeWidth={this.props.strokeWidth}
      />
    );
  }
}

export default Curve;
