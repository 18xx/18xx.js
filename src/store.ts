import { GameState } from './reducers/game';

export interface Store {
  getState: (hash: string) => Promise<GameState>;
  setState: (hash: string, body: GameState) => Promise<boolean>;
}
