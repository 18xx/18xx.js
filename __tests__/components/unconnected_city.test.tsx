import { List } from 'immutable';
import * as React from 'react';
import { Provider } from 'react-redux';
import * as renderer from 'react-test-renderer';

import Hexagon from '../../src/hexagon';

import { CityProps } from '../../src/components/city';
import UnconnectedCity from '../../src/components/unconnected_city';

import { stubStore } from '../support/store';

describe('UnconnectedCity', () => {
  const defaultProps: CityProps =  {
    hex: 'a1',
    hexagon: new Hexagon('east-west'),
    homeTokens: List(),
    mapDef: {
      companies: [],
      invertHexes: false,
      orientation: 'east-west',
      tileManifest: [],
    },
    num: 1,
    tokenState: List(),
  };

  describe('render', () => {
    describe('when there is one city', () => {
      const subject: React.ReactElement<UnconnectedCity> = (
        <Provider store={stubStore}>
          <UnconnectedCity {...defaultProps} />
        </Provider>
      );

      it('returns the correct result', () => {
        expect(renderer.create(subject)).toMatchSnapshot();
      });
    });

    describe('when there are two cities', () => {
      const subject: React.ReactElement<UnconnectedCity> = (
        <Provider store={stubStore}>
          <UnconnectedCity {...defaultProps} num={2} />
        </Provider>
      );

      it('returns the correct result', () => {
        expect(renderer.create(subject)).toMatchSnapshot();
      });
    });

    describe('when there three cities', () => {
      const subject: React.ReactElement<UnconnectedCity> = (
        <Provider store={stubStore}>
          <UnconnectedCity {...defaultProps} num={3} />
        </Provider>
      );

      it('returns the correct result', () => {
        expect(() => renderer.create(subject)).toThrow();
      });
    });
  });
});
