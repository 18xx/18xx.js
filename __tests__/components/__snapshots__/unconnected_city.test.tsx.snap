// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnconnectedCity render when there are two cities returns the correct result 1`] = `
<g>
  <svg
    onContextMenu={[Function]}
    x={15.42562584220407}
    y={34}
  >
    <circle
      className="city-circle"
      cx={20}
      cy={20}
      fill="white"
      r={19}
      stroke="#000"
      strokeWidth={2}
    />
  </svg>
  <svg
    onContextMenu={[Function]}
    x={55.42562584220407}
    y={54}
  >
    <circle
      className="city-circle"
      cx={20}
      cy={20}
      fill="white"
      r={19}
      stroke="#000"
      strokeWidth={2}
    />
  </svg>
</g>
`;

exports[`UnconnectedCity render when there is one city returns the correct result 1`] = `
<g>
  <svg
    onContextMenu={[Function]}
    x={35.42562584220407}
    y={44}
  >
    <circle
      className="city-circle"
      cx={20}
      cy={20}
      fill="white"
      r={19}
      stroke="#000"
      strokeWidth={2}
    />
  </svg>
</g>
`;
